lru.c

```diff
@@ -409,9 +409,46 @@ LRU_setdefault(LRU *self, PyObject *args)
 
     if (lru_ass_sub(self, key, default_obj) != 0)
         return NULL;
+
+    Py_INCREF(default_obj);
     return default_obj;
 }
 
+static PyObject *
+LRU_pop(LRU *self, PyObject *args)
+{
+    PyObject *key;
+    PyObject *default_obj = NULL;
+    PyObject *result;
+
+    if (!PyArg_ParseTuple(args, "O|O", &key, &default_obj))
+        return NULL;
+
+    result = lru_subscript(self, key);
+
+    if (result) {
+        /* Save and restore Python error indicator around deleting item --
+         * because between the last call of item getter (lru_subscript) and
+         * next call of item setter (lru_ass_sub), the item may have been
+         * removed from the LRU as the effect of arbitrary action in code
+         * elsewhere, thus incurring an error that will not be cleared by us.
+         */
+
+        /* Question: does this matter in the current implementation? */
+        PyObject *e_type, *e_value, *e_traceback;
+
+        PyErr_Fetch(&e_type, &e_value, &e_traceback);
+        lru_ass_sub(self, key, NULL);
+	PyErr_Restore(e_type, e_value, e_traceback);
+    } else if (default_obj) {
+        PyErr_Clear();
+        Py_INCREF(default_obj);
+        result = default_obj;
+    }
+
+    return result;
```

Comment:
I guess `dict.pop` throws a `KeyError` if default is not specified and key is missing. Better to keep same behavior here.

Reply:
You're right, it's in the standard library specs:

> If _key_ is in the dictionary, remove it and return its value, else return _default_. If _default_ is not given and _key_ is not in the dictionary, a `KeyError` is raised.

I'll be working on it shortly.

Reply:
I made a mistake reading my own code... The quoted (desired) behaviour is already the current behaviour, and this is tested by the following lines in the test script:

https://github.com/congma/lru-dict/blob/98ec21fcc6fb3722e476aa1b363bb936e2a011ba/test/test_lru.py#L248-L250